exchanges/stream/reporting.go

```diff
@@ -0,0 +1,119 @@
+	// Report logs the processing time for a received data packet and updates metrics.
+	// read is the time the data was read from the connection.
+	// data is the raw data received from the connection.
+	// err is any error that occurred while processing the data.
+	Report(read time.Time, data []byte, err error)
+	// Close closes the process reporter and handles any cleanup.
+	Close()
+}
+
+// SetProcessReportManager sets the ProcessReporterManager for the Websocket instance which will be used to create new ProcessReporter instances.
+// This will track metrics for processing websocket data.
+func (w *Websocket) SetProcessReportManager(m ProcessReporterManager) {
+	w.m.Lock()
+	defer w.m.Unlock()
+	w.processReporter = m
+}
+
+// NewDefaultProcessReporterManager returns a new defaultProcessReporterManager instance
+func NewDefaultProcessReporterManager() ProcessReporterManager {
+	return defaultProcessReporterManager{period: time.Minute}
+}
+
+// defaultProcessReporterManager is a default implementation of ProcessReporter
+type defaultProcessReporterManager struct{ period time.Duration }
+
+// New returns a new DefaultProcessReporter instance for a connection
+func (d defaultProcessReporterManager) New(conn Connection) ProcessReporter {
+	reporter := &defaultProcessReporter{ch: make(chan struct{})}
+	go reporter.collectMetrics(conn, d.period)
+	return reporter
+}
+
+// DefaultProcessReporter provides a thread-safe implementation of the ProcessReporter interface.
+// It tracks operation metrics, including the number of operations, average processing time, and peak processing time.
+type defaultProcessReporter struct {
+	operations          int64
+	errors              int64
+	totalProcessingTime time.Duration
+	peakProcessingTime  time.Duration
+	peakCause           []byte
+	ch                  chan struct{}
+	m                   sync.Mutex
+}
+
+// Report logs the processing time for a received data packet and updates metrics.
+func (r *defaultProcessReporter) Report(read time.Time, data []byte, err error) {
+	processingDuration := time.Since(read)
+	r.m.Lock()
+	defer r.m.Unlock()
+	r.operations++
+	if err != nil {
+		r.errors++
+	}
+	r.totalProcessingTime += processingDuration
+	if processingDuration > r.peakProcessingTime {
+		r.peakProcessingTime = processingDuration
+		r.peakCause = data
+	}
+}
+
+// Close closes the process reporter
+func (r *defaultProcessReporter) Close() {
+	r.m.Lock()
+	close(r.ch)
+	r.m.Unlock()
+}
+
+// collectMetrics runs in a separate goroutine to periodically log aggregated metrics.
+func (r *defaultProcessReporter) collectMetrics(conn Connection, period time.Duration) {
+	if period == 0 {
+		panic("period duration for collecting metrics must be greater than 0")
+	}
+	timer := time.NewTimer(time.Until(time.Now().Truncate(period).Add(period)))
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-r.ch:
+			return
+		case <-timer.C:
+			timer.Reset(time.Until(time.Now().Truncate(period).Add(period)))
```

Comment:
I think it would be better practice to reset the timer after the operation is done. Not that it should really impact with this kind of code especially the default time wait of 1 minute.

Reply:
[Done](https://github.com/thrasher-corp/gocryptotrader/pull/1734/commits/e9cbcd334557ae216ea399207be5ec7f185fadef)